File: python/1674-minimum-moves-to-make-array-complementary.py
# The idea is from Hao :) he is smart
class Solution:
    def minMoves(self, nums: List[int], limit: int) -> int:
        l = [0] * (2 + limit*2)
        N = len(nums)
        
        for i in range(N//2):
            min_num = min(nums[i], nums[N-1-i]) + 1
            max_num = max(nums[i], nums[N-1-i]) + limit
            total = nums[i] + nums[N-1-i]
            
            if min_num < total:
                if min_num>2: 
                    l[2]+=2
                    l[min_num] -=1
                else:
                    l[2]+=1
                l[total]-=1
            l[total+1]+=1
            l[max_num+1]+=1
        
        ans = float("infinity")
        curr = 0
        for i in range(2, len(l)):
            curr+=l[i]
            ans = min(ans, curr)
        
        return ans

            
            

        

File: hao/c++/1674-minimum-moves-to-make-array-complementary.cpp
class Solution {
public:
    int minMoves(vector<int>& nums, int limit) {
        
        unsigned int size = nums.size();
        unsigned int* nodes = (unsigned int*)malloc(6*size); // sizeof(int)*3/2
        // Loop the nums vector
        unsigned int j = 0;
        for (unsigned short i = 0; i < size/2; i++) {
            nodes[j++] = min(nums[i], nums[size-1-i]) + 1;
            nodes[j++] = max(nums[i], nums[size-1-i]) + limit;
            nodes[j++] = nums[i] + nums[size-1-i];
        }
        /*for (unsigned short i = 0; i < size*3/2; i++) {
            cout << nodes[i] << endl;
        }*/
        unsigned int i = 0;
        unsigned int max = 0;
        for (unsigned short i = 0; i < size*3/2; i++) {
            unsigned sum = 0;
            for (j = 0; j < 3*size/2; j+=3) {
                if (nodes[i] == nodes[j+2]) {
                    sum += 2;
                } else if (nodes[j] <= nodes[i] && nodes[i] <= nodes[j+1]) {
                    sum += 1;
                }
            }
            if (sum > max) {
                max = sum;
            }
        }
        return size - max;
    }
};